AWS: Update ObjectStorageLocationProvider hash to optimize for S3 performance

ookumuso opened the pull request:
The S3 team is introducing S3LocationProvider, a replacement to ObjectStorageLocationProvider that is better suited for the performance of Iceberg workloads running on Amazon S3. This configuration applies to Iceberg workloads using General Purpose and [Directory Buckets](https://docs.aws.amazon.com/AmazonS3/latest/userguide/directory-buckets-overview.html) in Amazon S3, and we expect it to further improve throughput of Iceberg workloads. Although this change can benefit all Iceberg workloads, the degree of improvement may vary based on specific workload characteristics.

This implementation changes the hash scheme from 32-bit base64 to 20-bit base2. The reduced character range allows S3 to automatically scale request capacity more quickly to match the demands of the target workload and reduce the amount of time that workloads observe throttle responses. A 20-bit hash allows for 2^20 possible prefixes.
This implementation also changes the path structure to reduce the number of directories created. The partition directories are eliminated, and the position of the hash is moved to before the data filename.

amogh-jahagirdar reviewed (commented):
Left some comments but a had a possibly naive question to just check my understanding:

In the past for object storage provider, we've used a wider character set in the hash portion of the file as a means to maximize entropy and ultimately improve heat distribution (https://github.com/apache/iceberg/pull/7128). With this new approach are we saying we can get a good enough heat distribution and at the same time enable s3 to scale capacity more quickly?

ookumuso commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  private static String dataLocation(Map<String, String> properties, String tableLocation) {
+    String dataLocation = properties.get(TableProperties.WRITE_DATA_LOCATION);
+    if (dataLocation == null) {
+      dataLocation = String.format("%s/data", tableLocation);
+    }
+    return dataLocation;
```
Doesn't nicely fit into a single line but it is easier to read I think, will update this.

ookumuso commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java:
```diff
@@ -0,0 +1,92 @@
+
+  private String computeHash(String fileName) {
+    HashCode hashCode = HASH_FUNC.hashString(fileName, StandardCharsets.UTF_8);
+    int hash = hashCode.asInt();
```
Works, updating

ookumuso commented:
> Left some comments but a had a possibly naive question to just check my understanding:
> 
> In the past for object storage provider, we've used a wider character set in the hash portion of the file as a means to maximize entropy and ultimately improve heat distribution (#7128). With this new approach are we saying we can get a good enough heat distribution and at the same time enable s3 to scale capacity more quickly?

@amogh-jahagirdar 
Yes, essentially both base64 and base2 distributes the traffic evenly which is the most crucial thing. With auto-scaling they can both support high TPS workloads. Main difference between them is how fast they auto-scale to the point where there are no more throttles observed. Base2 being more effective there due to every next char in the hash taking ~50% of the traffic.

danielcweeks commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java:
```diff
@@ -0,0 +1,92 @@
+ *   <li><code>tableLocation + "/data"</code>
+ * </ol>
+ *
+ * The data file is placed immediately under the data location. Partition names are <b>not</b>
```
I'm concerned about not including an option for partition values.  I generally agree that we don't need them and it's better for the overall key size to exclude that info, but I feel like it should still be an option and shouldn't impact the partition distribution.

ookumuso replied:
Intention of this change is to not only provide an optimal performance for S3 General Purpose Buckets but also for S3 Directory Buckets. This change essentially minimizes the directories required in the Directory Buckets, which have directories as a first-class citizen, whereas General Purpose buckets have a flat key space.
 
S3 General Purpose Buckets performance best practices suggest using entropy in prefixes to distribute load which the base2 achieves here but since Directory Buckets internally manage load distribution, creating sparse directories with that level of entropy can cause requests to be slower. So we wanted to provide a single entry point for all S3 relevant optimizations that captures these 2 bucket types.

danielcweeks replied:
@ookumuso it's helpful to understand the context behind why you may not want include the partition structure in the path, but I still think it should be an option (even if suboptimal).

jackye1995 replied:
Sounds like this can be just configurable, with a property like `write.location-provider.omit-partition-values-in-path`? We have seen people depending business workflows on partition path values which should be strongly discouraged, so I think it is better to default to not having partition structure in path. @danielcweeks what do you think?

danielcweeks commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java:
```diff
@@ -0,0 +1,92 @@
+ * s3://my-bucket/my-table/data/011101101010001111101000-00000-0-5affc076-96a4-48f2-9cd2-d5efbc9f0c94-00001.parquet
+ * </code>.
+ */
+public class S3LocationProvider implements LocationProvider {
```
Why not just include this in the ObjectStoreLocationProvider as an option to choose between base2 and base32?  It seems like we could include most of this in just the `computeHash` function.  That might also allow for choosing whether to include partition context.

ookumuso replied:
That was our original plan but with the partition values and other trailing directories removed, it turns into a very S3 specific location provider due to reasons mentioned in the above comment. I agree that if it was just the base2 optimization then it would fit into the ObjectStoreLocationProvider with one extra table property probably.

danielcweeks commented:
@ookumuso Overall, this looks like a great feature if this is better for S3 to repartition and distribute data, but it also seems like it would fit cleanly into the existing ObjectStoreLocationProvider as opposed to a separate provider.

danielcweeks commented on aws/src/main/java/org/apache/iceberg/aws/s3/S3LocationProvider.java:
```diff
@@ -0,0 +1,92 @@
+ * </ol>
+ *
+ * The data file is placed immediately under the data location. Partition names are <b>not</b>
+ * included. The data filename is prefixed with a 24-character binary hash, which ensures that files
```
The `24-character` prefix and lack of a directly marker (`/`) is a little problematic for older clients (like HadoopFS) and procedures like orphan files because anything that needs to list that directory will run into problems with the key space.  We already have an issue with this with the current provider, but this would compound that problem since there's no hierarch. 

This would make it a little easier to brake up prefix-separated ranges for listing (especially due to the low character limit), but that's something would need to look into on the orphan files action.

ookumuso replied:
Is the problem coming from the delimited list or is it some type parallelization issue due to lack of directories? I was thinking not having the directory essentially remove the requirement for discovering them and it can be just a flat list to discover everything under the write.data.path.

danielcweeks replied:
The issue is having too much cardinality for FS implementations that think in terms of directory structure.  By breaking this up we can address both scenarios (files system based and object store listing) at the same time.

ookumuso commented:
> @ookumuso Overall, this looks like a great feature if this is better for S3 to repartition and distribute data, but it also seems like it would fit cleanly into the existing ObjectStoreLocationProvider as opposed to a separate provider.

Thanks for the review Daniel! See my comment [here](https://github.com/apache/iceberg/pull/11112#discussion_r1763671169) regarding why we chose to go with a separate a provider. It was essentially done to capture optimizations for all bucket types in one place without having an impact on the existing default ObjectStoreLocationProvider.

ookumuso commented:
> > @ookumuso Overall, this looks like a great feature if this is better for S3 to repartition and distribute data, but it also seems like it would fit cleanly into the existing ObjectStoreLocationProvider as opposed to a separate provider.
> 
> Thanks for the review Daniel! See my comment [here](https://github.com/apache/iceberg/pull/11112#discussion_r1763671169) regarding why we chose to go with a separate a provider. It was essentially done to capture optimizations for all bucket types in one place without having an impact on the existing default ObjectStoreLocationProvider.

@danielcweeks let me know what you think about this

danielcweeks commented:
> @danielcweeks let me know what you think about this. One alternative is to maybe provide both as in having S3LocationProvider as is and a base2 option for the ObjectStoreLocationProvider to keep the option for partition values along with directories

@ookumuso I'm still of the opinion that we should just incorporate this into the existing object store location provider.  Ultimately, we can change the hashing behavior as the existing behavior is based on earlier discussions with S3, so replacing it or providing alternative options is fine. 

I do think partition values in the path still needs to be an option.

I think we also need to consider adding a path separator to help with some of the maintenance routines.  For example `s3://bucket/<prefix>/<db>/<table>/data/00101001/0010100111011...`. The `/` in the hash portion allows for `2^8` prefixes to allow listing under those paths for clients that list.  If we omit that, we get `2^24` prefixes that need to be listed under the data directory, which makes maintenance difficult.

ookumuso commented:
> > @danielcweeks let me know what you think about this. One alternative is to maybe provide both as in having S3LocationProvider as is and a base2 option for the ObjectStoreLocationProvider to keep the option for partition values along with directories
> 
> @ookumuso I'm still of the opinion that we should just incorporate this into the existing object store location provider. Ultimately, we can change the hashing behavior as the existing behavior is based on earlier discussions with S3, so replacing it or providing alternative options is fine.
> 
> I do think partition values in the path still needs to be an option.
> 
> I think we also need to consider adding a path separator to help with some of the maintenance routines. For example `s3://bucket/<prefix>/<db>/<table>/data/00101001/0010100111011...`. The `/` in the hash portion allows for `2^8` prefixes to allow listing under those paths for clients that list. If we omit that, we get `2^24` prefixes that need to be listed under the data directory, which makes maintenance difficult.

@danielcweeks Understood thanks for the feedback Daniel. Looks like it is not going to be feasible for us to remove `/` delimeter. With this, I agree that it doesn't make sense to offer a separate provider. I think it would be better if we update the existing provider to use base2 directly instead of opt in which can simplify the onboarding process for users as they can see the auto-scaling improvement right away. @jackye1995 Are you also okay with switching ObjectStoreLocationProvider to use base2 as a default as well? I can potentially cut down 24 bits to 20 or even 16 to reduce the length a bit.

For partition values, I can probably send a separate follow up as an option to remove them from the file name with a new table property so callers can decide but planning to exclude that for now.

jackye1995 commented:
Sorry for the late review, was busy with some internal work...

> it also seems like it would fit cleanly into the existing ObjectStoreLocationProvider as opposed to a separate provider.

+1 for that. I think the main concern was that this seems to be too S3 specific, but if there is no major concern from others I am good with changing the ObjectStoreLocationProvider directly, that actually helps adoption.

>  think we also need to consider adding a path separator to help with some of the maintenance routines.

I reviewed the internal experiments related to this, and having no path separators as well as having at least 24 bits I believe are all important based on load testing. If the concern is compatibility with other systems, would it make sense to only use this approach for locations that begins with `s3://` or `s3a://`?

ookumuso commented:
@danielcweeks @amogh-jahagirdar @jackye1995 As discussed above, made the change to update the existing ObjectStoreLocationProvider to use base2 entropy and also added new config to omit the partition values. I set it to false as default to make it backwards compatible but let me know whether you prefer that to be true to prevent customers to rely on partition values in the file path.

I kept the "/" delimiter in place due to concerns raised by @danielcweeks but would like to understand what would it take to remove that as a separate thread to potentially follow up in a separate PR. [Orphan](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java#L311) clean up path was one thing identified but I am curious whether there are any other things we need to consider to make it a viable change

ookumuso commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -123,6 +123,11 @@ static class ObjectStoreLocationProvider implements LocationProvider {
       } else {
         this.context = pathContext(tableLocation);
       }
+      this.excludePartitionValues =
```
Updated

jackye1995 commented:
Any additional comments? @amogh-jahagirdar @nastra @danielcweeks

danielcweeks commented on docs/docs/configuration.md:
```diff
@@ -77,6 +77,7 @@ Iceberg tables support table properties to configure table behavior, like the de
 | write.metadata.previous-versions-max                 | 100                         | The max number of previous version metadata files to keep before deleting after commit                                                                                                            |
 | write.spark.fanout.enabled                           | false                       | Enables the fanout writer in Spark that does not require data to be clustered; uses more memory                                                                                                   |
 | write.object-storage.enabled                         | false                       | Enables the object storage location provider that adds a hash component to file paths                                                                                                             |
+| write.object-storage.omit-partition-values           | false                       | Excludes the partition values from the file path                                                                                                                                                  |
```
This property introduces a "double-negative".  I think we should invert this to be something like `write.object-storage.partitioned-paths = true`

ookumuso replied:
Updated

danielcweeks commented on core/src/main/java/org/apache/iceberg/TableProperties.java:
```diff
@@ -244,6 +244,11 @@ private TableProperties() {}
   public static final boolean OBJECT_STORE_ENABLED_DEFAULT = false;
 
+  // Excludes the partition values in the path when set to true and object store is enabled
+  public static final String WRITE_OBJECT_STORE_OMIT_PARTITION_VALUES =
```
Also mentioned in the docs comment, but we should invert this.

danielcweeks commented:
> looks good to me. I think there are 2 pending topics and I am approving since I am good with the current way it is written:
> 
> 1. should we do it for all paths, or just limit to `s3://` and `s3a://` paths: I think given that even the documentation of this is in AWS, it is probably fine to just apply this strategy for all users using object storage location provider.

I think we should treat all of the prefixes the same (e.g. `s3`,`s3n`,`s3a`).  These are all just aliases to map different implementations to, but the behavior should be the same.

> 2. should we add / in the middle of the binary hash values: that would mitigate the directory listing issue, but will impact the ability for scaling up S3 due to the additional slashes in the middle. My understanding is that the directory listing concern could be resolved if we move all HDFS-based listing to use prefix-based listing, and the team can take that as a follow up item.

I think we should add a single `/` to break up the prefix as I noted in [my comment above](https://github.com/apache/iceberg/pull/11112#issuecomment-2371796750).  I doubt that a single slash is going to have a major impact on s3's ability to partition effectively.  My understanding is that S3 doesn't treat the `/` character any different than any other in determining heat/partitioning and since it's a single static value, it would probably be skipped over.

That slash is important for directory based file system implementations as omitting it would result in listing effectively the full contents of the table and would likely break clients because there are too many results to hold in memory.

jackye1995 commented:
> That slash is important for directory based file system implementations as omitting it would result in listing effectively the full contents of the table and would likely break clients because there are too many results to hold in memory.

when you say "directory based file system implementations", do you mean implementation of catalog, or FileIO? 

And what is the use case of this? I thought the only place we are doing something like this is in orphan file removal, but for that we are just iterating through the directories to find matching files, we are holding all the matching files in memory, not holding everything in memory.

jackye1995 commented:
> The hashing is all performed at the top-level of the data directory. That means to list that directory using a FS based FileIO, you need to list that directory which could have 2^24 subdirs.

I see. But I think this is already a problem today with the 64^6 (2^36) subdirs, and it's actually reducing the number of directories for new tables?

I looked at the code of orphan files removal more closely, looks like the issue is at [here](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java#L360-L366), where it forces listing the entire directory and then check for if the sub directory count exceeds the maximum.

And to properly fix this, my understanding is that we should switch to use prefix-based listing, so that if the FileIO implements `SupportsPrefixListing`, then that API should be used rather than HDFS FileSystem. Would you agree this is the right way to fix this specific issue? I think it's better than trying to fit S3 into HDFS use pattern. @danielcweeks

jackye1995 commented:
> Adding a slash within the hash allows that to be broken up and listed in a distributed way to build the dataset used by spark for the full listing.

Btw, another way to solve the problem, just to brainstorm here:

My understanding from S3 side (@ookumuso correct me if I am wrong) is that the ideal case that also works better with the new S3 directory bucket is that: all the files are in the same directory, we don't even want `/` by the end of the hash prefix. So files will be with stored in places like `/data/0101010101010101-file-name.parquet`

Would this be a better strategy? If we store files like this, it also does not create the too many nested directory problem during orphan file cleanup that you described. For the code pointer above:

```java
      for (FileStatus file : fs.listStatus(path, pathFilter)) {
        if (file.isDirectory()) {
          subDirs.add(file.getPath().toString());
        } else if (file.isFile() && predicate.test(file)) {
          matchingFiles.add(file.getPath().toString());
        }
      }

      // stop listing if the number of direct sub dirs is bigger than maxDirectSubDirs
      if (subDirs.size() > maxDirectSubDirs) {
        remainingSubDirs.addAll(subDirs);
        return;
      }
```

This will just go through all files and store the matching files in a single call because there is no nested directories under `/data/`, and I don't see any risk of memory issue (assuming matching files don't blow up memory)

danielcweeks commented:
> > The hashing is all performed at the top-level of the data directory. That means to list that directory using a FS based FileIO, you need to list that directory which could have 2^24 subdirs.
> 
> I see. But I think this is already a problem today with the 64^6 (2^36) subdirs, and it's actually reducing the number of sub-directories with the new strategy?

It's a problem with both strategies.  If we weren't changing the hashing, we should introduce a path separator (though likely at a different level) for the current strategy.  Since we're considering changing this, we should fix this as well.

> I looked at the code of orphan files removal more closely, looks like the issue is at [here](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java#L360-L366), where it forces listing the entire directory and then check for if the sub directory count exceeds the maximum.

My understanding is that the current approach does an initial top-level listing of the data dir and then farms subsequent listings out to executors.  This would allow for parallel listing in a more reasonable way without running into limits.  I believe this might already work with the default depths (though we should double-check that).

> And to properly fix this, my understanding is that we should switch to use prefix-based listing, so that if the FileIO implements `SupportsPrefixListing`, then that API should be used rather than HDFS FileSystem. Would you agree this is the right way to fix this specific issue? I think it's better than trying to fit S3 into HDFS use pattern. @danielcweeks

I think we need to try to accommodate both since there's a fair amount of usage of S3A.  I believe we can do a much better job by splitting the prefix range up for parallel listings with  any FileIO that supports prefix listing, but the current implementation relies on S3A and is built around that behavior.  My point is we can accommodate both reasonably, but I agree there's a better solution with prefix listing.

jackye1995 commented:
> My understanding is that the current approach does an initial top-level listing of the data dir and then farms subsequent listings out to executors. This would allow for parallel listing in a more reasonable way without running into limits. I believe this might already work with the default depths (though we should double-check that).

I think [the code](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java#L360-L372) does not do what you describe, please correct me if I am wrong or looking at the wrong place:

```java
      for (FileStatus file : fs.listStatus(path, pathFilter)) {
        if (file.isDirectory()) {
          subDirs.add(file.getPath().toString());
        } else if (file.isFile() && predicate.test(file)) {
          matchingFiles.add(file.getPath().toString());
        }
      }

      // stop listing if the number of direct sub dirs is bigger than maxDirectSubDirs
      if (subDirs.size() > maxDirectSubDirs) {
        remainingSubDirs.addAll(subDirs);
        return;
      }
```

Because it must first finish doing the `fs.listStatus` call to produce the full list of `subDirs` before checking against `maxDirectSubDirs`, you will end up listing all the random prefixes before able to terminate the call, and this happens before farming out subsequent calls to any executor.

And the files in a table is listed based on [this code](https://github.com/apache/iceberg/blob/main/spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/DeleteOrphanFilesSparkAction.java#L311-L319):

```java
    // list at most MAX_DRIVER_LISTING_DEPTH levels and only dirs that have
    // less than MAX_DRIVER_LISTING_DIRECT_SUB_DIRS direct sub dirs on the driver
    listDirRecursively(
        location,
        predicate,
        hadoopConf.value(),
        MAX_DRIVER_LISTING_DEPTH,
        MAX_DRIVER_LISTING_DIRECT_SUB_DIRS,
        subDirs,
        pathFilter,
        matchingFiles);
```

which starts listing at table root location level and list up to depth 3, that will cover listing the `/data/` directory on driver since that is only 1 level under the root location. And I don't think adjusting the listing depth or listing direct dub dirs would fix the issue. But overall I think we are all agreeing that that we have a problem today about doing orphan file removal using S3A with object storage layout.

Since we all agree that there is a better path forward using prefix-listing, and the S3 team is planning to work on that, should we still add the slash in the middle of the hash? It feels like a pretty hacky fix. I will let @ookumuso and @drewschleit chime in here, what I heard was that having the middle slash would impact throttling handling, and is also a bad practice for directory buckets as it creates more sparse directories. Even if we want to do the middle slash fix, I think we should add it with a config which is disabled by default, and focus on the right way to fix it.

ookumuso commented:
> Since we all agree that there is a better path forward using prefix-listing, and the S3 team is planning to work on that, should we still add the slash in the middle of the hash? It feels like a pretty hacky fix. I will let @ookumuso and @drewschleit chime in here, what I heard was that having the middle slash would impact throttling handling, and is also a bad practice for directory buckets as it creates more sparse directories. Even if we want to do the middle slash fix, I think we should add it with a config which is disabled by default, and focus on the right way to fix it.

For S3 General Purpose bucket scaling, we mainly care about the entropy and not the directories so where "/" is doesn't really matter that much. This is a concern for directory buckets due to sparse directories as in too many directories with very little number of files under each since directories are first class citizens there and there is no prefix based scaling concern. 

Moving forward with the base2 change even as is with the "/" in the end is crucial today for S3 performance so I would like to tackle these concerns separately. It looks like with the base2 change we are removing a bunch of work from the driver since we are reducing directory count from 64^6 (2^36) to 2^20 which is a significant drop. 

Prefix listing is a good idea but I think it might be limited in its effectiveness against directory buckets since there is no native api to list any prefix there. See the comment [here](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html) particularly ` Directory buckets - For directory buckets, only prefixes that end in a delimiter (/) are supported.`

Given there are complications there, would like get your take on moving forward with this unless you think it makes things worse since this is important for S3 auto-scaling in general.

jackye1995 commented:
> we mainly care about the entropy and not the directories so where "/" is doesn't really matter that much.

> Moving forward with the base2 change even as is with the "/" in the end is crucial today for S3 performance so I would like to tackle these concerns separately.

@ookumuso thanks for the quick response, just to confirm we are on the same page with these comments, what @danielcweeks is suggesting is to have the "/" also in the middle, e.g. "/data/01010101/01010101/file-name.parquet". It is not just in the end as you said later, e.g. "/data/0101010101010101/file-name.parquet". 

For the specific suggestion to have the "/" also in the middle, does it matter or not matter in the entropy design?

ookumuso commented:
> > we mainly care about the entropy and not the directories so where "/" is doesn't really matter that much.
> 
> > Moving forward with the base2 change even as is with the "/" in the end is crucial today for S3 performance so I would like to tackle these concerns separately.
> 
> @ookumuso thanks for the quick response, just to confirm we are on the same page with these comments, what @danielcweeks is suggesting is to have the "/" also in the middle, e.g. "/data/01010101/01010101/file-name.parquet". It is not just in the end as you said later, e.g. "/data/0101010101010101/file-name.parquet".
> 
> For the specific suggestion to have the "/" also in the middle, does it matter or not matter in the entropy design?

Yeah that does not matter for entropy design, auto-scaling behavior will be same on general purpose buckets. For directory buckets, I guess it is the same amount of sparseness since it does not change the total directory count. I just agree that it looks a bit hacky overall

jackye1995 commented:
Okay in that case at least I feel much more comfortable with adding slashes in the middle, thanks for confirming.

Looking at the orphan file delete logic, If we really want to fully take advantage of how the recursive + distributed file listing is implemented today, it seems like the right way to do the hashing path is to add not just 1 slash, but multiple slashes, something like "/data/0101/0101/0101/0101/..../file-name.parquet", so it gives enough directory depth out of the box, and that everything further nested is sent to executors for further listing. What do you think? @danielcweeks 

And would that still be fine with S3 scaling, or a single slash is better? @ookumuso

ookumuso commented:
> And would that still be fine with S3 scaling, or a single slash is better? @ookumuso

Probably not but I will dig in a bit more and confirm this tomorrow just to be sure.

> Looking at the orphan file delete logic, If we really want to fully take advantage of how the recursive + distributed file listing is implemented today, it seems like the right way to do the hashing path is to add not just 1 slash, but multiple slashes, something like "/data/0101/0101/0101/0101/..../file-name.parquet", so it gives enough directory depth out of the box, and that everything further nested is sent to executors for further listing. What do you think? @danielcweeks

Is there an optimal number of directories and depth? maybe we can just create those and put rest of the entropy into the fileName. For example: `/data/010/001/100/01010101001-file-name.parquet`. This can reduce both the sparse directory problem and help with orphan clean-up?

jackye1995 commented:
> Is there an optimal number of directories and depth?

The hard-coded default is 3. Usually, there is a /data/ folder which takes 1 level of depth, but that might not always be the case. So if we stick to that default, then we should ensure 3 levels of depth.

> put rest of the entropy into the fileName

When `write.object-storage.partitioned-path=true`, that will create paths like `/data/010/001/100/01010101001-key=val/file-name.parquet`, which will be a bit awkward.

We should potentially introduce a separator concept `write.object-storage.separator` that defaults to slash, and can be other values like `-` and can produce paths like `/data/010/001/100/01010101001-key=val-file-name.parquet`, which seems like the most ideal solution.

ookumuso commented:
What do you think about this approach @danielcweeks:

> Is there an optimal number of directories and depth? maybe we can just create those and put rest of the entropy into the fileName. For example: /data/010/001/100/01010101001-file-name.parquet. This can reduce both the sparse directory problem and help with orphan clean-up?

It might be a nice win-win case both solving sparse directory problem and orphan clean up. As @jackye1995 it is a bit weird for partitioned-paths but we can do it for not partitioned paths only, so it would like:
* partitioned-path=true:  `/data/010/001/100/01010101001/key=val/file-name.parquet`
* partitioned-path=false: `/data/010/001/100/01010101001-file-name.parquet`

danielcweeks commented:
> What do you think about this approach @danielcweeks:
> 
> > Is there an optimal number of directories and depth? maybe we can just create those and put rest of the entropy into the fileName. For example: /data/010/001/100/01010101001-file-name.parquet. This can reduce both the sparse directory problem and help with orphan clean-up?
> 
> It might be a nice win-win case both solving sparse directory problem and orphan clean up. As @jackye1995 it is a bit weird for partitioned-paths but we can do it for not partitioned paths only, so it would like:
> 
> * partitioned-path=true:  `/data/010/001/100/01010101001/key=val/file-name.parquet`
> * partitioned-path=false: `/data/010/001/100/01010101001-file-name.parquet`

I like removing additional path when `partitioned-path=false` (that's a great idea).

I'm just wondering about the where we want to put the slashes in the bit field.  Breaking it up by three bits means that each recursive listing just operates on 8 subpaths, which feels too small.  I feel like we might want to move to four bits (i.e. `/data/0100/0110/0010/101010010110`).  This leaves the leaf paths with a reasonable `2^12` paths and each parent with 16, which feels like enough to parallelize at some level.

ookumuso commented:
Sounds good @danielcweeks, I will work on this change and update the PR!

ookumuso commented:
@danielcweeks @jackye1995 Updated the change to divide entropy into dirs so we follow the following format now:

    partitioned-path=true: /data/0100/0110/0010/10101001/key=val/file-name.parquet
    partitioned-path=false: /data/0100/0110/0010/10101001-file-name.parquet

Set the dir depth to 3 and dir length to 4 as discussed.

jackye1995 commented on docs/docs/configuration.md:
```diff
@@ -39,52 +39,53 @@ Iceberg tables support table properties to configure table behavior, like the de
 
 ### Write properties
 
-| Property                                             | Default                     | Description                                                                                                                                                                                       |
```
looks like this is auto-formatted by IDE, can we minimizes the changes to only the new configs? If there is not a good way, I guess we will have to commit the change like this.

jackye1995 commented:
@danielcweeks any further comments?

danielcweeks commented:
@ookumuso a couple small remaining comments:

1. It looks like we're only using 20 bits, not 24 like in the description
2. The aws.md docs don't reflect the updated pathing structure

ookumuso commented:
> @ookumuso a couple small remaining comments:
> 
>     1. It looks like we're only using 20 bits, not 24 like in the description
> 
>     2. The aws.md docs don't reflect the updated pathing structure

@danielcweeks thanks:
1. Intentionally dropped the bits as it should more than enough, updated the PR description and aws.doc as well
2. Updated the aws.md doc with the new path structure

ookumuso commented:
Any additional concerns @danielcweeks ?

danielcweeks commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -172,10 +193,45 @@ private static String pathContext(String tableLocation) {
+     * @return 1001/1001/1001/10011001 with depth 3 and length 4
+     */
+    private String dirsFromHash(String hash) {
+      if (hash.length() < ENTROPY_DIR_DEPTH * ENTROPY_DIR_LENGTH) {
```
We should use a `Preconditions.check` here.

ookumuso replied:
Yeah maybe I am a bit over-paranoid here, given I have the right unit tests to cover any changes and length is 20 bits as final string(after addressing the other feedback), maybe will just remove this condition?

danielcweeks replied:
If it's all internal to this class, I think we can safely assume that `Integer.toBinaryString(hashCode.asInt() | Integer.MIN_VALUE)` will produce the correct number of bits.  The check is seems unnecessary since it's not user supplied inputs.

ookumuso replied:
Done

danielcweeks commented on core/src/main/java/org/apache/iceberg/LocationProviders.java:
```diff
@@ -108,10 +108,15 @@ public String newDataLocation(String filename) {
-    private static final BaseEncoding BASE64_ENCODER = BaseEncoding.base64Url().omitPadding();
-    private static final ThreadLocal<byte[]> TEMP = ThreadLocal.withInitial(() -> new byte[4]);
+    // the starting index of the lower 20-bits of a 32-bit binary string
+    private static final int HASH_BINARY_STRING_START_INDEX = 12;
```
Can we flip this to be to be `HASH_BINARY_STRING_BITS = 20`, since that's more applicable to what we're constructing?  The start index is confusing because it only has meaning in context to the length of what we're skipping from the bit field.

ookumuso replied:
Done

danielcweeks reviewed (approved):
Just waiting on checks.  Thanks @ookumuso !

jackye1995 reviewed (approved):
Also looks good to me! Thanks for the work!

jackye1995 commented:
looks like CI has passed, merging. Thanks for the work and patience @ookumuso , and thanks for the review @danielcweeks !
